src/promptflow/promptflow/_utils/tool_utils.py

```diff
@@ -305,6 +306,33 @@ def load_function_from_function_path(func_path: str):
         )
 
 
+def assign_tool_input_index_for_ux_order_if_needed(tool):
+    tool_type = tool.get("type")
+    if should_preserve_tool_inputs_order(tool_type) and "inputs" in tool:
+        inputs_dict = tool["inputs"]
+        input_index = 0
+        for input_name, settings in inputs_dict.items():
+            # 'uionly_hidden' indicates that the inputs are not the tool's inputs.
+            # They are not displayed on the main interface but appear in a popup window.
+            # These inputs are passed to UX as a list, maintaining the same order as generated by func parameters.
+            # Skip the 'uionly_hidden' input type because the 'ui_hints: index' is not needed.
+            if "input_type" in settings.keys() and settings["input_type"] == "uionly_hidden":
+                continue
+            settings.setdefault(UI_HINTS, {})
+            settings[UI_HINTS]["index"] = input_index
+            input_index += 1
+
+
+def should_preserve_tool_inputs_order(tool_type):
+    """
+    Currently, we only automatically add input indexes for the custom_llm tool,
+    following the order specified in the tool interface or YAML.
+    This is because, as of now, only the custom_llm tool has such a requirement.
```

Comment:
We may add more detailed explanation about "only the custom_llm tool has such a requirement"

Reply:
Added, thanks.